### DISCUSSION TITLE
PlutoUI -Recompute button (reactive trigger)

### DISCUSSION THREAD

**Discussion by casasgomezuribarri** [fonsp/Pluto.jl#2423] [Questions about using Pluto] [answered]
I love this functionality!

But I would love there to be a similar button for "Compute" rather than "Recompute" (i.e. the cell does not evaluate by default. It does evaluate every time the button is clicked, but the first evaluation occurs after the first click)

EDIT: I want it to allow downloading plots only when they look exactly like the user wants. The download button does not work for me (error message below), so fixing that would also solve my case. However, I do think that having cells that only evaluate on command would be very cool (especially while building a notebook).

This is my error message: 

> Failed to show value:
> MethodError: no method matching write(::Base64.Base64EncodePipe, ::Plots.Plot{Plots.PlotlyJSBackend})
> Closest candidates are:
> write(::IO, ::Any) at io.jl:672
> write(::IO, ::Any, !Matched::Any...) at io.jl:673
> write(!Matched::FileIO.Stream, ::Any...)

And this is my cell script:

> DownloadButton(p, "plot.png")

Changing the backend did not solve it. Should I wrap my plot inside something? I was thinking something like `png(p)`, although that doesn't work either:

> MethodError: no method matching splitdir(::Plots.Plot{Plots.PlotlyJSBackend})
> Closest candidates are:
> splitdir(!Matched::FilePathsBase.AbstractPath) at ~/.julia/packages/FilePathsBase/9kSEl/src/path.jl:314
> splitdir(!Matched::String) at path.jl:130
> splitdir(!Matched::AbstractString) at path.jl:581

**Comment by fonsp** [comment]
You could do something like:

```julia
@bind should_compute Checkbox()
```

```julia
if should_compute
    do_thing()
end
```

or:

```julia
@bind clicked CounterButton()
```

```julia
if clicked > 0
    do_thing()
end
```

**Comment by fonsp** [comment, marked as answer]
About the download button: it currently only supports a `String` or a `Vector{UInt8}` (byte array) as input. You can render a plot to png data like so:

```julia
repr(MIME("image/png"), p)
```
so you button would look like:
```julia
DownloadButton(repr(MIME("image/png"), p), "plot.png")
```

**Comment by casasgomezuribarri** [comment]
Thanks! Didn't think about MIME... That is a great and neat solution :D 

The problem with using `CounterButton()` and `Checkbox()` is that once the condition is met, any further changes to the plot would trigger a new download (right?), which is not ideal. 

For the sake of context, the notebook uses `PlutoUI` to facilitate the choice of hyperparameters for an ML model, and plots the results as a confusion matrix. Only the confusion matrix of the model that works is interesting, and that is the one that will be downloaded... 

I'd considered the question answered, thanks!

**Reply by Pangoraw** [reply]
>The problem with using CounterButton() and Checkbox() is that once the condition is met, any further changes to the plot would trigger a new download (right?), which is not ideal.

The `DownloadButton` shows a button that needs to be clicked in order to trigger a download. So once the user has achieved the plot result they want, they can just click it.

You can see what it looks like on this PlutoUI demo:
https://featured.plutojl.org/basic/plutoui.jl#f74f434a-c768-11ea-079c-fb707e6ba17b
